Remove expensify-common library from project

fabioh8010 opened the pull request:
### Details

This PR extracts the two methods from `expensify-common` library that are being used in this project, and remove `expensify-common` as it's not necessary anymore.

### Related Issues

https://github.com/Expensify/react-native-onyx/issues/260

### Automated Tests

Not applicable as this PR is only extracting some methods from `expensify-common`.

### Linked PRs

Not applicable.

fabioh8010 commented:
I have read the CLA Document and I hereby sign the CLA

amyevans commented on lib/Str.js:
```diff
@@ -0,0 +1,30 @@
+    },
+
+    /**
+     * Checks if parameter is a string or function
+     * if it is a function then we will call it with
+     * any additional arguments.
```
```suggestion
     * Checks if parameter is a string or function.
     * If it is a function, then we will call it with
     * any additional arguments.
```

amyevans commented on lib/Str.js:
```diff
@@ -0,0 +1,30 @@
+     * any additional arguments.
+     *
+     * @param {String|Function} parameter
+     * @returns {String}
```
This doesn't necessarily return a string, can we update it?

hayata-suenaga replied:
I think if a function is passed to the first argument (instead of string), that function should return a string.

I don't know how to provide type information of callbacks, but hopefully when we migrate onyx, we can type that.

roryabraham replied:
This function is pretty weird 🙃

fabioh8010 replied:
This function returns `string` if we pass a `string` to `parameter`, or returns the result of `parameter` if it's a function. So, I will change to `{*}` as if we pass a function it can return any type. WDYT?

fabioh8010 replied:
@hayata-suenaga In this [example](https://github.com/Expensify/react-native-onyx/blob/main/lib/withOnyx.js#L121) the function is returning a `boolean` instead of a `string`.

hayata-suenaga replied:
I just don't know why this function exists 😭

amyevans reviewed (commented):
Some nit picks but otherwise 👍

neil-marcellini reviewed (dismissed):
Looks good! Why do we need to do this? Is it because expensify-common is going to be using Typescript now? We are only adding type declarations there so I'm still unsure 🤔 

cc @hayata-suenaga

hayata-suenaga commented:
@neil-marcellini you're right. We don't necessarily have to do this.

However, while examining `onyx`, we discovered there are only two functions used from `expensify-common` in `react-native-onyx` and we decided just remove the library completely.

Eventually, we will migrate `onyx` to TypeScript (vs. just creating definition files). When we do that, it makes things earlier if we have one less internal library.

Relevant slack discussion: https://expensify.slack.com/archives/C0593T50UHG/p1684971334416899?thread_ts=1684958481.278379&cid=C0593T50UHG

Relevant design doc section: https://docs.google.com/document/d/1LNcgYeYyovQ88gNjLIgBqBReLtbLAbxeAbna9QKxeAQ/edit#bookmark=id.f26etjel5h0v

roryabraham commented on lib/Str.js:
```diff
@@ -0,0 +1,30 @@
+    startsWith(haystack, needle) {
+        return _.isString(haystack)
+                && _.isString(needle)
+                && haystack.substring(0, needle.length) === needle;
```
NAB:

```suggestion
                && haystack.startsWith(needle);
```

roryabraham reviewed (approved):
LGTM

roryabraham commented:
Nothing glaring here but I think there's enough little things that I'll wait for @fabioh8010 to make a few of the requested changes before merging. If anyone else is feeling 🤠 feel free to merge.

fabioh8010 commented:
I've addressed the comments and improved `result` function's JSDoc to be more clearer. @hayata-suenaga As I just copied these two functions from `expensify-common`, should we address those changes there as well in the PR Blazej created?

hayata-suenaga commented:
>  should we address those changes there as well in the PR Blazej created?

which PR are you talking about???
